### PR TITLE
KEYCLOAK-11877 Theme support for ValidateX509CertificateUsername

### PR DESCRIPTION
**Pull request by slaskawi**
[KEYCLOAK-11877](https://issues.redhat.com/browse/KEYCLOAK-11877)

This Pull Request introduces theme support for `ValidateX509CertificateUsername` authenticator

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 1]
```diff
@@ -48,7 +50,7 @@ public void authenticate(AuthenticationFlowContext context) {
         if (certs == null || certs.length == 0) {
             logger.debug("[ValidateX509CertificateUsername:authenticate] x509 client certificate is not available for mutual SSL.");
             context.getEvent().error(Errors.USER_NOT_FOUND);
-            Response challengeResponse = errorResponse(Response.Status.UNAUTHORIZED.getStatusCode(), "invalid_request", "X509 client certificate is missing.");
+            Response challengeResponse = ErrorPage.error(context.getSession(), context.getAuthenticationSession(), Response.Status.UNAUTHORIZED, Messages.X509_CERTIFICATE_MISSING);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: When would this error happen? It seems if there is no certificate x509 authentication should not be attempted?

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 2]
```diff
@@ -63,7 +65,7 @@ public void authenticate(AuthenticationFlowContext context) {
         if (config == null) {
             logger.warn("[ValidateX509CertificateUsername:authenticate] x509 Client Certificate Authentication configuration is not available.");
             context.getEvent().error(Errors.USER_NOT_FOUND);
-            Response challengeResponse = errorResponse(Response.Status.UNAUTHORIZED.getStatusCode(), "invalid_request", "Configuration is missing.");
+            Response challengeResponse = ErrorPage.error(context.getSession(), context.getAuthenticationSession(), Response.Status.UNAUTHORIZED, Messages.MISSING_CONFIGURATION);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: Missing configuration should return a 500 internal server error. Message returned to user should be generic, it should not expose internal error messages to end-users.

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 3]
```diff
@@ -63,7 +65,7 @@ public void authenticate(AuthenticationFlowContext context) {
         if (config == null) {
             logger.warn("[ValidateX509CertificateUsername:authenticate] x509 Client Certificate Authentication configuration is not available.");
             context.getEvent().error(Errors.USER_NOT_FOUND);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: This looks wrong. Config is missing, but it is showing that the user is not found

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 4]
```diff
@@ -48,7 +50,7 @@ public void authenticate(AuthenticationFlowContext context) {
         if (certs == null || certs.length == 0) {
             logger.debug("[ValidateX509CertificateUsername:authenticate] x509 client certificate is not available for mutual SSL.");
             context.getEvent().error(Errors.USER_NOT_FOUND);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: This is not right, certificate is not found, but it's showing user is not found

File: services/src/main/java/org/keycloak/services/messages/Messages.java [hunk 5]
```diff
@@ -147,6 +151,8 @@
 
     public static final String IDENTITY_PROVIDER_AUTHENTICATION_FAILED = "identityProviderAuthenticationFailedMessage";
 
+    public static final String AUTHENTICATION_FAILED = "AuthenticationFailedMessage";
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/messages/Messages.java]: authenticationFailedMessage

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 6]
```diff
@@ -76,8 +78,7 @@ public void authenticate(AuthenticationFlowContext context) {
                     .validateExtendedKeyUsage();
         } catch(Exception e) {
             logger.error(e.getMessage(), e);
-            // TODO use specific locale to load error messages
-            Response challengeResponse = errorResponse(Response.Status.UNAUTHORIZED.getStatusCode(), "invalid_request", e.getMessage());
+            Response challengeResponse = ErrorPage.error(context.getSession(), context.getAuthenticationSession(), Response.Status.UNAUTHORIZED, Messages.AUTHENTICATION_FAILED, e.getMessage());
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: We should just use internalServerError, with 500 as the return status code here I think

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 7]
```diff
@@ -86,9 +87,7 @@ public void authenticate(AuthenticationFlowContext context) {
         if (userIdentity == null) {
             context.getEvent().error(Errors.INVALID_USER_CREDENTIALS);
             logger.errorf("[ValidateX509CertificateUsername:authenticate] Unable to extract user identity from certificate.");
-            // TODO use specific locale to load error messages
-            String errorMessage = "Unable to extract user identity from specified certificate";
-            Response challengeResponse = errorResponse(Response.Status.UNAUTHORIZED.getStatusCode(), "invalid_request", errorMessage);
+            Response challengeResponse = ErrorPage.error(context.getSession(), context.getAuthenticationSession(), Response.Status.UNAUTHORIZED, Messages.X509_USER_IDENTITY_EXTRACT_ERROR);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: internalServerError, with 500 I think

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 8]
```diff
@@ -100,36 +99,34 @@ public void authenticate(AuthenticationFlowContext context) {
         }
         catch(ModelDuplicateException e) {
             logger.modelDuplicateException(e);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: Not sure why there would be a ModelDuplicateException. There's no writes to the DB are there?!

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java [hunk 9]
```diff
@@ -100,36 +99,34 @@ public void authenticate(AuthenticationFlowContext context) {
...
-            Response challengeResponse = errorResponse(Response.Status.UNAUTHORIZED.getStatusCode(), "invalid_request", errorMessage);
+            Response challengeResponse = ErrorPage.error(context.getSession(), context.getAuthenticationSession(), Response.Status.UNAUTHORIZED, Messages.AUTHENTICATION_FAILED, e.getMessage());
             context.failure(AuthenticationFlowError.INVALID_USER, challengeResponse);
             return;
         }
         catch(Exception e) {
             logger.error(e.getMessage(), e);
-            String errorMessage = String.format("X509 certificate authentication's failed. Reason: \"%s\"", e.getMessage());
-            Response challengeResponse = errorResponse(Response.Status.UNAUTHORIZED.getStatusCode(), "invalid_request", errorMessage);
+            Response challengeResponse = ErrorPage.error(context.getSession(), context.getAuthenticationSession(), Response.Status.UNAUTHORIZED, Messages.AUTHENTICATION_FAILED, e.getMessage());
```
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: internalServerError, with 500 I think

File: services/src/main/java/org/keycloak/services/messages/Messages.java [hunk 10]
```diff
@@ -58,6 +60,8 @@
 
     public static final String MISSING_TOTP = "missingTotpMessage";
 
+    public static final String MISSING_CONFIGURATION = "missingConfigurationMessage";
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/messages/Messages.java]: Remove this. We should return details like this to end-users.

File: services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java (hunk 8 above)
Comment (stianst) [reply services/src/main/java/org/keycloak/authentication/authenticators/x509/ValidateX509CertificateUsername.java]: If it's valid catch, then 500 with internalServerError

File: services/src/main/java/org/keycloak/services/messages/Messages.java (hunk 5 above)
Comment (stianst) [inline services/src/main/java/org/keycloak/services/messages/Messages.java]: Remove this. We should return details like this to end-users.

File: services/src/main/java/org/keycloak/services/messages/Messages.java [hunk 11]
```diff
@@ -238,4 +244,6 @@
     public static final String DELEGATION_FAILED = "delegationFailedMessage";
     public static final String DELEGATION_FAILED_HEADER = "delegationFailedHeader";
 
+    public static final String X509_CERTIFICATE_MISSING = "authenticatorX509CertificateMissing";
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/messages/Messages.java]: Remove this. We should return details like this to end-users.

File: services/src/main/java/org/keycloak/services/messages/Messages.java [hunk 12]
```diff
@@ -238,4 +244,6 @@
...
+    public static final String X509_USER_IDENTITY_EXTRACT_ERROR = "authenticatorX509UserIdentityExtractError";
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/messages/Messages.java]: Remove this. We should return details like this to end-users.

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties [hunk 13]
```diff
@@ -247,12 +247,14 @@ couldNotObtainTokenMessage=Could not obtain token from identity provider.
 unexpectedErrorRetrievingTokenMessage=Unexpected error when retrieving token from identity provider.
 unexpectedErrorHandlingResponseMessage=Unexpected error when handling response from identity provider.
 identityProviderAuthenticationFailedMessage=Authentication failed. Could not authenticate with identity provider.
+authenticationFailedMessage=Authentication failed. Reason: {0}
```
Comment (stianst) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties]: Remove this. We should return details like this to end-users.

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties [hunk 14]
```diff
@@ -266,6 +268,8 @@ identityProviderAlreadyLinkedMessage=Federated identity returned by {0} is alrea
 confirmAccountLinking=Confirm linking the account {0} of identity provider {1} with your account.
 confirmEmailAddressVerification=Confirm validity of e-mail address {0}.
 confirmExecutionOfActions=Perform the following action(s)
+authenticatorX509CertificateMissing=X509 client certificate is missing.
```
Comment (stianst) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties]: Remove this. We should return details like this to end-users.

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties [hunk 15]
```diff
@@ -266,6 +268,8 @@ identityProviderAlreadyLinkedMessage=Federated identity returned by {0} is alrea
...
+authenticatorX509UserIdentityExtractError=Unable to extract user identity from specified certificate.
```
Comment (stianst) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties]: Remove this. We should return details like this to end-users.

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties [hunk 16]
```diff
@@ -287,6 +291,7 @@ locale_sv=Svenska
 
 backToApplication=&laquo; Back to Application
 missingParameterMessage=Missing parameters\: {0}
+missingConfigurationMessage=Missing configuration
```
Comment (stianst) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties]: Remove this. We should return details like this to end-users.

**Review by stianst — CHANGES_REQUESTED** [review]
See inline comments. In general this fixes the bug reported in KEYCLOAK-11877, but reviewing this I found a general issue with the wrong level of details being returned to the user, which we should address.

Also, since we didn't catch that it was returning a JSON error instead of an error page it is clear that there's a complete lack of testing here. We need to at least add some tests for error conditions like this.

**Comment by slaskawi** [comment]
> See inline comments. In general this fixes the bug reported in KEYCLOAK-11877, but reviewing this I found a general issue with the wrong level of details being returned to the user, which we should address.

@stianst Ok, no problem, I'll fix them.

> Also, since we didn't catch that it was returning a JSON error instead of an error page it is clear that there's a complete lack of testing here. We need to at least add some tests for error conditions like this.

@stianst The tests seem to be there - `X509DirectGrantTest`. Moreover, they seem not to be executed in Travis (I'll check why) and I broke them with this PR. Let me fix this as well.

**Comment by slaskawi** [comment]
After investigation, together with @stianst we decided to close this one. See comments in [KEYCLOAK-11877](https://issues.redhat.com/browse/KEYCLOAK-11877).

**Pull request closed without merging**